test_conformance/api/test_api_consistency.cpp

```diff
@@ -473,14 +437,689 @@ int test_consistency_pipes(cl_device_id deviceID, cl_context context,
+            "returned an empty set");
+    }
+
+    return TEST_PASS;
+}
+
+int test_consistency_2d_image_from_buffer(cl_device_id deviceID,
+                                          cl_context context,
+                                          cl_command_queue queue,
+                                          int num_elements)
+{
+    // clGetDeviceInfo, passing CL_DEVICE_IMAGE_PITCH_ALIGNMENT or
+    // CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT
+    // May return 0, indicating that device does not support Creating a 2D Image
+    // from a Buffer.
+    cl_int error;
+
+    const cl_image_format imageFormat = { CL_RGBA, CL_UNORM_INT8 };
+    const size_t imageDim = 2;
+    const size_t elementSize = 4;
+    const size_t bufferSize = imageDim * imageDim * elementSize;
+
+    clMemWrapper buffer;
+    clMemWrapper image;
+
+    cl_uint imagePitchAlignment = 0;
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_IMAGE_PITCH_ALIGNMENT,
+                            sizeof(imagePitchAlignment), &imagePitchAlignment,
+                            NULL);
+    test_error(error,
+               "Unable to query "
+               "CL_DEVICE_IMAGE_PITCH_ALIGNMENT");
+
+    cl_uint imageBaseAddressAlignment = 0;
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT,
+                            sizeof(imageBaseAddressAlignment),
+                            &imageBaseAddressAlignment, NULL);
+    test_error(error,
+               "Unable to query "
+               "CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT");
+
+    bool supports_cl_khr_image2d_from_buffer =
+        is_extension_available(deviceID, "cl_khr_image2d_from_buffer");
+
+    if (imagePitchAlignment == 0 || imageBaseAddressAlignment == 0)
+    {
+        // This probably means that Creating a 2D Image from a Buffer is not
+        // supported.
+
+        // Test setup:
+        buffer =
+            clCreateBuffer(context, CL_MEM_READ_ONLY, bufferSize, NULL, &error);
+        test_error(error, "Unable to create test buffer");
+
+        // Check that both queries return zero:
+        test_assert_error(imagePitchAlignment == 0,
+                          "CL_DEVICE_IMAGE_PITCH_ALIGNMENT returned a non-zero "
+                          "value but CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT "
+                          "returned 0");
+        test_assert_error(imagePitchAlignment == 0,
+                          "CL_DEVICE_IMAGE_PITCH_ALIGNMENT returned 0 but "
+                          "CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT returned a "
+                          "non-zero value");
+
+        // clGetDeviceInfo, passing CL_DEVICE_EXTENSIONS
+        // Will not describe support for the cl_khr_image2d_from_buffer
+        // extension if device does not support Creating a 2D Image from a
+        // Buffer.
+        test_assert_error(supports_cl_khr_image2d_from_buffer == false,
+                          "Device does not support Creating a 2D Image from a "
```

Comment:
Hmm, it's the presence of the extension that defines whether the feature is supported. I would change the message to say something along the lines of `cl_khr_image2d_from_buffer` is supported but an invalid value was returned for ..."

Reply:
Following the logic in the spec appendix, the feature is supported / not supported based on the device queries, and the extension must be reported consistent with these queries.  I think this is six of one, half-dozen of the other though, and it really doesn't matter which one gets checked first, so long as everything is consistent.  Are any consistency checks missing?

Reply:
Well the spec appendix is not normative and my reading of the normative section says otherwise. Both `CL_DEVICE_IMAGE_PITCH_ALIGNMENT` and `CL_DEVICE_IMAGE_BASE_ADDRESS_ALIGNMENT` are specified to return 0 when the feature is not supported but only the text for `CL_DEVICE_EXTENSIONS` defines a logical equivalence between the presence of the extension string and support for the feature.

> Are any consistency checks missing?

Not AFAICT. The test will be correct once the comment on the assert above is fixed.

Since this is mostly a readability issue and not a terrible one, I'll resolve this.